File: alexandria/ui/src/io/intino/alexandria/ui/displays/components/NumberEditable.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;

public class NumberEditable<B extends Box> extends AbstractNumberEditable<B> {
	private double value;
	private double min;
	private double max;
	private double step;

    public NumberEditable(B box) {
        super(box);
    }

	public double value() {
		return value;
	}

	public NumberEditable value(double value) {
		this.value = value;
		return this;
	}

	public double min() {
		return min;
	}

	public NumberEditable min(double min) {
		this.min = min;
		return this;
	}

	public double max() {
		return max;
	}

	public NumberEditable max(double max) {
		this.max = max;
		return this;
	}

	public double step() {
		return step;
	}

	public NumberEditable step(double step) {
		this.step = step;
		return this;
	}

	public void update(double value) {
		value(value);
		notifier.refresh(value);
	}

	public void notifyChange(Double value) {
	}

}

File: alexandria/ui-elements/src/displays/components/NumberEditable.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractNumberEditable from "../../../gen/displays/components/AbstractNumberEditable";
import NumberEditableNotifier from "../../../gen/displays/notifiers/NumberEditableNotifier";
import NumberEditableRequester from "../../../gen/displays/requesters/NumberEditableRequester";
import TextField from '@material-ui/core/TextField';
import InputAdornment from '@material-ui/core/InputAdornment';

const styles = theme => ({
	default : {
		width: "100%"
	}
});

class NumberEditable extends AbstractNumberEditable {
	state = {
		value : ""
	};

	constructor(props) {
		super(props);
		this.notifier = new NumberEditableNotifier(this);
		this.requester = new NumberEditableRequester(this);
	};

	handleChange(e) {
		const value = e.target.value;
		if (value === "") return;
		this.requester.notifyChange(value);
		this.setState({ value: value });
	};

	render() {
		const { classes } = this.props;
		const value = this.state.value !== "" ? this.state.value : (this.props.min !== -1 ? this.props.min : 0);
		const label = this.props.label !== "" ? this.props.label : undefined;
		const format = this.props.format != null && this.props.format !== "default" ? this.props.format.split(" ")[0] : "body1";

		return (
			<TextField format={format} style={this.style()} className={classes.default} label={label} type="number"
					   value={value} onChange={this.handleChange.bind(this)}
					   inputProps={{
						   min: this.props.min !== -1 ? this.props.min : undefined,
						   max: this.props.max !== -1 ? this.props.max : undefined,
						   step: this.props.step !== -1 ? this.props.step : undefined
					   }}
					   InputProps={{
						   startAdornment: this.props.prefix !== undefined ? <InputAdornment position="start">{this.props.prefix}</InputAdornment> : undefined,
						   endAdornment: this.props.suffix !== undefined ? <InputAdornment position="end">{this.props.suffix}</InputAdornment> : undefined
					   }}></TextField>
		);
	};

	refresh = (value) => {
		this.setState({ "value": value });
	};

}

export default withStyles(styles, { withTheme: true })(NumberEditable);